File: src/lib/crdt/gcounter.ts
export default class GCounter {
    private _value: number;

    constructor(value = 0) {
        this._value = value;
    }

    get value() {
        return this._value;
    }

    inc(v = 1) {
        return (this._value += v);
    }

    merge(other: GCounter) {
        return (this._value = other._value);
    }
}

File: server/src/main/kotlin/pt/up/fe/sdle/crdt/GCounter.kt
package pt.up.fe.sdle.crdt

class GCounter(
    private var value: Int = 0,
) {
    fun getValue() = value

    fun inc(v: Int = 1): Int {
        if (v < 0) throw IllegalArgumentException("Cannot decrement a GCounter")

        value += v
        return value
    }

    fun merge(other: GCounter): Int {
        value = maxOf(value, other.value)
        return value
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as GCounter

        return value == other.value
    }

    override fun hashCode(): Int {
        return value
    }

    override fun toString(): String = "GCounter($value)"
}
